docs: Add Python examples to docs (#1047)
TimefoldAI/timefold-solver: docs/src/modules/ROOT/pages/constraints-and-score/score-calculation.adoc

:sectnums:
:icons: font

> Using Java's Streams API or Python's generator expressions,

Reviewer: This is another example where maybe talking about a particular language doesn't make sense anymore.

This can be "Let's implement ... using a functional approach:".
Then inside the tab for a particular language, we can add more details specific to that language, if we need to.

we could implement an xref:constraints-and-score/score-calculation.adoc#easyJavaScoreCalculation[easy score calculator]
that uses a functional approach:

[tabs]
====
Java::
+
[source,java,options="nowrap"]
----
    private int doNotAssignAnn() {
        return schedule.getShifts().stream()
                .filter(Shift::isEmployeeAnn)
                .mapToInt(shift -> -1)
                .sum();
    }
----

Python::
+
[source,python,options="nowrap"]
----
def do_not_assign_ann() -> int:
    return sum(-1 for shift in schedule.shifts if shift.employee == 'Ann');
----
====

However, that scales poorly because it doesn't do an xref:constraints-and-score/performance.adoc#incrementalScoreCalculation[incremental calculation]:
When the planning variable of a single `Shift` changes, to recalculate the score,
the normal Streams API has to execute the entire stream from scratch.

[#constraintStreams]
== Introducing Constraint Streams

> Constraint streams are a Functional Programming form of incremental score calculation in plain Java, Python or Kotlin

Reviewer: ```suggestion
Constraint streams are a functional programming form of incremental score calculation
```

that is easy to read, write and debug.
The API should feel familiar if you're familiar with Java Streams or SQL.

> The Constraint Streams API enables you to write similar code in pure Java, Python or Kotlin,

Reviewer: ```suggestion
The Constraint Streams API enables you to write similar code in a language of your choice,
```

[#constraintStreamsConfiguration]
== Creating a constraint stream

> To use the Constraint Streams API in your project, first write a pure Java, Python or Kotlin `ConstraintProvider` implementation similar

Reviewer: ```suggestion
To use the Constraint Streams API in your project, first write a `ConstraintProvider` implementation similar
```

The following code snippet first groups all visits by the vehicle that will be used,
and averages all the service durations using the `ConstraintCollectors.average(...)` collector.

[tabs]
====
Java::
+
[source,java,options="nowrap"]
----
private Constraint averageServiceDuration(ConstraintFactory constraintFactory) {
    return constraintFactory.forEach(Visit.class)
            .groupBy(Visit::getVehicle, average(Visit::getServiceDuration))
            .penalize(HardSoftScore.ONE_SOFT,

>                     (vehicle, averageServiceDuration) -> (int) averageServiceDuration)

Reviewer: Personally, I'd prefer `HardSoftLongScore` and removing the cast.
Simpler code for the example.

Author: `averageServiceDuration` is a double, not a long. Cast is needed regardless.

    vehicleA.visits.append(visit1)
    visit2 = Visit("3", "Paul", LOCATION_3, 40, tomorrow_08_00, tomorrow_10_00, timedelta(minutes=30))
    vehicleA.visits.append(visit2)

    (constraint_verifier.verify_that(vehicle_capacity)
                        .given(vehicleA, visit1, visit2)
                        .penalizes_by(20))
----
====

This test ensures that, if a vehicle capacity is exceeded, that the constraint penalizes accordingly.
The following line creates a shared `ConstraintVerifier` instance and initializes the instance with the
`VehicleRoutingConstraintProvider`:

[tabs]
====
Java::
+
[source,java,options="nowrap"]
----
    private ConstraintVerifier<VehicleRoutingConstraintProvider, VehicleRoutePlan> constraintVerifier
            = ConstraintVerifier.build(new VehicleRoutingConstraintProvider(), VehicleRoutePlan.class, Vehicle.class, Visit.class);
----

Python::
+
[source,python,options="nowrap"]
----
constraint_verifier = ConstraintVerifier.build(vehicle_routing_constraints, VehicleRoutePlan, Vehicle, Visit)
----
====

The `@Test` annotation indicates that the method is a unit test in a testing framework of your choice.

> Constraint Verifier works with many testing frameworks including JUnit, AssertJ and PyTest.

Reviewer: ```suggestion
Constraint Verifier works with testing frameworks popular in your ecosystem, including JUnit and PyTest.
```

Timefold Solver supports two other types of score calculation.

[#easyJavaScoreCalculation]

> === Easy Java/Python score calculation

Reviewer: Same comment as above; specifically naming the technology seems unnecessary. The heading, as well as the references, should change.

> An easy way to implement your score calculation in Java or Python.

Reviewer: ```suggestion
An easy way to implement your score calculation as imperative code:
```

* Advantages:

> ** Plain old Java/Python: no learning curve.

Reviewer: ```suggestion
** No learning curve; use the language you're already familiar with.
```

** Opportunity to delegate score calculation to an existing code base or legacy system.
** Useful for prototyping.
* Disadvantages:
** Slower, typically not suitable for production.
** Does not scale because there is no xref:constraints-and-score/performance.adoc#incrementalScoreCalculation[incremental score calculation].
** Can not xref:constraints-and-score/understanding-the-score.adoc[explain the score].

> To start using Easy Java/Python score calculation, implement the one method of the interface ``EasyScoreCalculator``:

Reviewer: ```suggestion
To start using Easy score calculator, implement the one method of the interface ``EasyScoreCalculator``:
```

Author: Unsure about that phasing; either "To start using an easy score calculator, ...", or "To start using easy score calculation, ...".

[#incrementalJavaScoreCalculation]

> === Incremental Java/Python score calculation

Reviewer: Dtto.

> A way to implement your score calculation incrementally in Java or Python.

Reviewer: ```suggestion
A way to implement your score calculation incrementally in a supported programming language.
```

** Hard to read
*** Regular score constraint changes can lead to high maintenance costs.

> To start using Incremental Java/Python score calculation,

Reviewer: ```suggestion
To start using Incremental score calculator,
```
